File: C++/726.cpp
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
public:
    string countOfAtoms(string formula) {
        if (formula.empty()) return formula;
		// used to calculate the count of parentheses
        vector<int> factors{1};
        int fact = 1;
		// number of element
        int num = 0;
		// element name
        string element = "";
		// which digit is this number
        int digit = 1;
		// counts
        map<string, int> cnts;
        for (int i = formula.size() - 1; i >= 0; --i) {
			// if we see a ')', we need to calculate the current count of the parentheses
            if (formula[i] == ')') {
			// if we don't have a numer, then the number is 1
                if (!num) num = 1;
                factors.push_back(num);
                fact *= num;
                num = 0;
                digit = 1;
			// '(' means we are going out one parentheses, lower the counts
            } else if (formula[i] == '(') {
                fact /= factors.back();
                factors.pop_back();
			// if it's a number, we can calculate the current number
            } else if (formula[i] >= '0' && formula[i] <= '9') {
                num += (formula[i] - '0') * digit;
                digit *= 10;
			// if it's a char, means it's part of a element
            } else if (formula[i] >= 'a' && formula[i] <= 'z') {
                element.push_back(formula[i]);
			// Upper case means we have finished reading an element
            } else if (formula[i] >= 'A' && formula[i] <= 'Z') {
				// set num to 1 if there is no number
                if (!num) num = 1;
				// element should be at most 2 chars, num is the current number, fact is the accumulate count of parentheses
                element = string(1, formula[i]) + element;
                cnts[element] += num * fact;
				// ready to record a new element
                element.clear();
				// ready to record a new number
                num = 0;
                digit = 1;
            }
        }
        string res = "";
		// write the result
        for (auto& p : cnts) res += p.first + ((p.second > 1) ? to_string(p.second) : "");
        return res;
    }
};
__________________________________________________________________________________________________
sample 9156 kb submission
class Solution {
public:
    int pos, n;
    string s;
    
    int parseNumber()
    {
        int ans = 0;
        while (pos < n && isdigit(s[pos]))
            ans = ans * 10 + s[pos++] - '0';
        return ans;
    }
    
    string parseName()
    {
        string ans = "";
        ans += s[pos++];
        while (pos < n && islower(s[pos]))
            ans += s[pos++];
        return ans;
    }
    
    string countOfAtoms(string formula) {
        vector<pair<string,int>> v;
        pos = 0;
        n = formula.size();
        s = formula;
        while (pos < n)
        {
            if (s[pos] == '(')
            {
                v.push_back({"(", 0});
                pos++;
            }
            else if (s[pos] == ')')
            {
                pos++;
                int k = 1;
                if (pos < n && isdigit(s[pos]))
                {
                    k = parseNumber();
                }
                for (int i = v.size() - 1; i >= 0; i--)
                {
                    if (v[i].first == "(")
                    {
                        v[i].first = ".";
                        break;
                    }
                    v[i].second *= k;
                }
            }
            else
            {
                string name = parseName();
                int k = 1;
                if (pos < n && isdigit(s[pos]))
                    k = parseNumber();
                v.emplace_back(name, k);
            }
        }
        
        map<string, int> cnt;
        for (auto p : v)
            if (p.first != ".")
                cnt[p.first] += p.second;
        string ans = "";
        for (auto p : cnt)
        {
            ans += p.first;
            if (p.second > 1)
            {
                stringstream ss;
                ss << p.second;
                ans += ss.str();
            }
        }
        
        return ans;
    }
};
__________________________________________________________________________________________________

File: Python3/726.py
__________________________________________________________________________________________________
sample 24 ms submission
class Solution:
    def countOfAtoms(self, formula: str) -> str:
        
        stack, atom, dic, count, coeff, c = [], '', collections.defaultdict(int), 0, 1, 0
        
        for i in formula[::-1]:
            if i.isdigit():
                count += int(i) * (10 ** c)
                c += 1
            elif i == ')':
                stack.append(count)
                coeff *= count
                count = c = 0
            elif i == '(':
                coeff //= stack.pop()
                count = c = 0
            elif i.isupper():
                atom += i
                dic[atom[::-1]] += (count or 1) * coeff
                atom = ''
                count = c = 0
            else:
                atom += i
        
        check = []
        for atom in dic:
            check.append((atom, dic[atom]))
        check.sort(key=lambda x:x[0])
        ans = ''
        for atom, count in check:
            ans += atom
            if count > 1:
                ans += str(count)
        return ans
__________________________________________________________________________________________________
sample 13188 kb submission
class Solution:
    def multiply(self,multiplier):
        product=1
        
        for c in multiplier:
            product = product * c
            
        return product
    
    def sort_answer(self,dic):
        output=[]
        
        for key,val in sorted(dic.items()):
            output.append(key)
            if val==1:
                continue
            output.append(str(val))
        
        return ''.join(output)
        
    def countOfAtoms(self, formula: str) -> str:
        from collections import defaultdict
        
        multAr = []
        parentheses=0
        d=defaultdict(int)
        element=None
        elems=[]
        num=None
        
        i=len(formula)-1
        
        while i>=0:
            if formula[i].isdigit():
                if i==len(formula)-1 or num==None:
                    num=formula[i]
                else:
                    num= formula[i]+num
                    
            if formula[i].islower():
                element = formula[i-1:i+1]
                if num==None:
                    d[element]+= self.multiply(multAr)
                else:
                    d[element]+=int(num) * self.multiply(multAr)
                num=None  
                i-=2
                elems.append(element)
                continue
            
            elif formula[i].isupper():
                element=formula[i]
                if num==None:
                    d[element]+= self.multiply(multAr)
                else:
                    d[element]+=int(num) * self.multiply(multAr)
                num=None
                elems.append(element)
                
            elif formula[i]==')':
                multAr.append(int(num))
                num=None
                
            elif formula[i]=='(':
                multAr.pop()
            
            i-=1
            
        return self.sort_answer(d)
__________________________________________________________________________________________________
